text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,211 @@
+    None || return      // returns from function    + and are the same to how boolean || works
+    Some(2) && Some(3)  // Some(3)
+    None && Some(1)     // None
+    Some(3) && None     // None
+
+    Some(2) || Some("hello") // Error: LogicalOr<Option<&str>> not implemented for Option<i32>
+    Some(2) || 2 || 3        // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+#### For `Result<T, E>`
+```rust
+struct MyError;
+
+fn foo() -> Result<i32, MyError> {
+    Ok(3)
+}
+
+fn main() {
+    Ok(4) || Ok(5);           // == Ok(4)
+    Err<MyError{}> || Ok(5);  // == Ok(5)
+    Ok(4) || foo();           // == Ok(4) (foo is *not* called)
+    Err<MyError{}> || foo();  // == Ok(3) (foo is called)
+    Err<MyError{}> || 3;      // == 3
+    Ok(2) || 1;               // == 2
+
+    Ok(2) || Ok("hello");  // Error: LogicalOr<Result<&str, _>> not implemented for Result<i32, _>
+    Ok(2) || 2 || 3;       // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+The feature should be thought about as moving the logic from methods and into the current function.
+It maps very seamlessly from using the methods and is equivalent in use without having to worry about
+the naming convention of the short circuit methods. The same mental state of short circuit from
+bools applies directly, without having any recourse to "truthiness" which is not a desirable trait.
+
+This RFC also proposes to deprecate the `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(||
+...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Option<T>` and `.or(...)`,
+`.or_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Result<T, E>` since
+using this feature renders them unneeded.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The basis of this proposal are the two new traits. These should be implemented along the same lines
+as other operator traits. However, the desugaring should follow the `match` output in the previous
+section so as to obtain the desired short circuit operation.
+
+Once the traits have been implemented, several trait implementations should be added to the std library and the methods marked as deprecated.
+
+```rust
+impl trait LogicalOr<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Option<T> {
+    type Output = T;
+}
+
+impl trait LogicalAnd<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<Result<T, E>> for Result<T, E> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Result<T, E> {
+    type Output = T;
+}
+
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Leaves the `||` and the `&&` as not strictly boolean operators, which might hurt readability
+2. Could lead to similarities to C++'s Operator bool() which => truthiness and is undesirable.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
Two more things I'd like to see discussed in here

- Why the method to allow combining the sides, vs just something like `-> Option<Rhs>`?

- Why two traits, vs using one method that splits into the two parts, one kept on `&&` and the other kept on `||`

Reply:
I don't understand the two traits question. Is it really necessary to discuss why each operator should have its own trait?